flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -1195,9 +1184,22 @@ private ApplicationReport startAppMaster(
         yarnClient.submitApplication(appContext);
 
         LOG.info("Waiting for the cluster to be allocated");
+
+        waitAppSubmission(yarnClient, appId, false);
+
+        // since deployment was successful, remove the hook
+        ShutdownHookUtil.removeShutdownHook(deploymentFailureHook, getClass().getSimpleName(), LOG);
+
+        return () -> waitAppSubmission(yarnClient, appId, true);
```

Comment:
Let's use concrete classes. This makes it a lot easier to test them properly.

Reply:
Got it.